### ISSUE TITLE
Isolating only pedestrian(1) and  people(2) in the Visdrone dataset

### ISSUE THREAD

**Issue by fatbringer** [ultralytics/ultralytics#6876]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to just isolate bounding boxes for people and pedestrians using the code below, which is modified from the yaml file found in the repo https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/VisDrone.yaml
```
    import os
from pathlib import Path
import yaml

from ultralytics.utils.downloads import download

def visdrone2yolo(dir):
    from PIL import Image
    from tqdm import tqdm

 
    def convert_box(size, box):
        # Convert VisDrone box to YOLO xywh box
        dw = 1. / size[0]
        dh = 1. / size[1]
        return (box[0] + box[2] / 2) * dw, (box[1] + box[3] / 2) * dh, box[2] * dw, box[3] * dh

    (dir / 'labels').mkdir(parents=True, exist_ok=True)  # make labels directory
    pbar = tqdm((dir / 'annotations').glob('*.txt'), desc=f'Converting {dir}')
    for f in pbar:
        img_size = Image.open((dir / 'images' / f.name).with_suffix('.jpg')).size
        lines = []
        with open(f, 'r') as file:  # read annotation.txt
            for row in [x.split(',') for x in file.read().strip().splitlines()]:
                if row[4] == '0':  # VisDrone 'ignored regions' class 0
                    continue
                #modifications to include only pedestrian and people
                if row[5] == "1" or "2":
                    #cls = int(row[5]) - 1
                    cls = 0 #turns all the people classes into 0, pls make sure consistent across all training
                    box = convert_box(img_size, tuple(map(int, row[:4])))
                    lines.append(f"{cls} {' '.join(f'{x:.6f}' for x in box)}\n")
                    with open(str(f).replace(f'{os.sep}annotations{os.sep}', f'{os.sep}labels{os.sep}'), 'w') as fl:
                        fl.writelines(lines)  # write label.txt
    
    print("done")


# Download

with open("VisDrone.yaml", "r") as stream:
    try:
        yamlfile = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

print("cwd is: ", Path.cwd())
base_dir = Path.cwd()
dir = base_dir.joinpath(yamlfile['path'])
#dir = ( Path.cwd() / yamlfile['path'] )  # dataset root dir
print("dataset root dir is: ", dir)
urls = ['https://github.com/ultralytics/yolov5/releases/download/v1.0/VisDrone2019-DET-train.zip',
        'https://github.com/ultralytics/yolov5/releases/download/v1.0/VisDrone2019-DET-val.zip',
        'https://github.com/ultralytics/yolov5/releases/download/v1.0/VisDrone2019-DET-test-dev.zip',
        'https://github.com/ultralytics/yolov5/releases/download/v1.0/VisDrone2019-DET-test-challenge.zip']

#download(urls, dir=dir, curl=True, threads=4)

# Convert
for d in 'VisDrone2019-DET-train', 'VisDrone2019-DET-val', 'VisDrone2019-DET-test-dev':
    visdrone2yolo(dir / d)  # convert VisDrone annotations to YOLO labels


 ```

my key modification is at the for loop portion to just get the relevant classes
However, i am still getting like buses and cars. 
Am i indicating the wrong classes? or is my if condition not working correctly?

**Comment by glenn-jocher** [comment]
@fatbringer hello! It looks like there may be a small issue with your if condition in the loop filtering for the pedestrian and people classes. You're currently using `if row[5] == "1" or "2":`, which will always evaluate to `True` because `"2"` is truthy in Python.

To fix this, you should change the condition to explicitly compare `row[5]` with both `"1"` and `"2"` like this:

```python
if row[5] == "1" or row[5] == "2":
```

or use the `in` operator with a tuple:

```python
if row[5] in ("1", "2"):
```

This correction should properly filter out only the pedestrians and people from your dataset according to the VisDrone class indices. Give it a try, and if you encounter any further issues, our documentation and community discussions are great places to seek additional assistance. 📘👥

**Comment by fatbringer** [comment]
@glenn-jocher thanks so much! That really does fix it

There's another question is, pedestrian and people classes doesnt really seem to include all of the people per se.
For example, in these images.


![0000003_00231_d_0000016_wbbox](https://github.com/ultralytics/ultralytics/assets/93758179/75f1239f-2930-4402-a0b0-73749a8698ae)
For example, there are alot of people in the basketball court in the background that are not bounded up.

![0000071_03522_d_0000006_wbbox](https://github.com/ultralytics/ultralytics/assets/93758179/95163605-0cae-4f86-b288-fd7db4920a82)
In this example, there are also alot of people who are not included in the pedestrian and people classes.

Would you know which classes I should also account for to adequately cover them?

**Comment by glenn-jocher** [comment]
@fatbringer I'm glad the previous fix worked for you! Regarding the additional people not being detected, this could be due to the way the VisDrone dataset is annotated. Not all instances may be labeled as pedestrians or people, especially if they are small, occluded, or far away.

For a comprehensive coverage, you can examine the VisDrone dataset's class annotations and include additional classes that may encompass people in different contexts (e.g., 'person sitting', 'child', etc.). It's also worthwhile to review the annotation guidelines for the dataset to understand the criteria used for different classes.

In the cases you've mentioned, manual verification of the dataset might be necessary to identify all relevant classes. Once you've determined which additional classes are needed, you can update your filtering code to include those classes as well. Keep in mind that datasets like VisDrone may have inherent limitations in terms of annotation consistency and completeness.

**Comment by fatbringer** [comment]
@glenn-jocher alright i understand. Just another question, would you or is there anyone else on the team familiar with the TinyPersons dataset? 

otherwise, thanks for your help! Have a good day :)

**Comment by glenn-jocher** [comment]
@fatbringer Yes, that's correct! Each label line follows the format of "class index, x_center, y_center, width, height" relative to the image size, with only those 5 values per line. I'm glad I could assist you, and have a great day too! 🙂

**Comment by fatbringer** [comment]
Hey there

I have beeen running the training on the Visdrone dataset and i got this set of results
![results](https://github.com/ultralytics/ultralytics/assets/93758179/0df9c506-1aac-42da-afbc-123465bc7143)

It's a little worrying that the mAP50 isnt doing too well. I also found it odd that the box_loss and cls_loss are still quite high.
My current settings are 300 epochs (it stopped at 248 epochs due to patience =50), batch size = 32 and imgsz 640.

Would you be able to share what else i could tune for better results?

**Comment by glenn-jocher** [comment]
Hello! Based on your results, if the mAP @fatbringer.5 is not satisfactory and the losses seem high, consider the following adjustments:

1. **Data Augmentation**: Enhance the generalization capability of the model by applying different data augmentation techniques.
2. **Learning Rate Schedule**: Alter the learning rate, or use a learning rate scheduler to adjust it during training.
3. **Model Complexity**: Try using a more complex model or adjusting the depth/multiplier of the current architecture.
4. **Class Balancing**: Address any class imbalance in your dataset through techniques like focal loss or class weights.
5. **Extended Training**: Sometimes simply training for more epochs can help, as long as you monitor for overfitting.
6. **Hyperparameters**: Experiment with different hyperparameter settings, potentially using an automated search process like Hyperband or Bayesian Optimization.

For more detailed guidance on these strategies, visit our documentation on model optimization and tuning. Good luck with improving your model's performance! 🚀

**Comment by bshakya77** [comment]
Hello,

I am training YOLO8n and YOLO8s model on VisDrone2019 dataset using Google Collab Pro. I have a few questions which are listed below:

1. Do we need to create VisDrone.xaml file separately or it gets automatically used when we provide the 'VisDrone.yaml' as data argument while training. My training command is as follows:

> model = YOLO('yolov8n.pt') 
> results = model.train(data='VisDrone.yaml', epochs=25, imgsz=224, batch=16)

2. There are very limited objects detect while training mostly the car objects are only detected. How can I make it detect other class labels in the same amount?

3. If I need to train 'yolov8-p2' model. Can I simply used the following command:

>model = YOLO('yolov8-p2.yaml')  
results = model.train(data='VisDrone.yaml', epochs=25, imgsz=224)

Does yolov8-p2 also have pretrained model like YOLO8n model?

Thank you.

Regards,
Bijay

**Comment by glenn-jocher** [comment]
@bshakya77 hello Bijay,

1. You should ensure the `VisDrone.yaml` is correctly set up and present in your working directory. The training process will not automatically create it, but will use the file provided in the `data` argument.

2. To improve detection across various classes, confirm that your dataset is balanced with respect to class representation. You might also explore different data augmentation techniques and tweak hyperparameters such as the class weights to encourage the model to learn under-represented categories.

3. For training a `yolov8-p2` model, you would typically load a pre-existing configuration or checkpoint. If there is a pre-trained `yolov8-p2` model available, you can specify it in the same manner as the `yolov8n.pt`:

```python
model = YOLO('yolov8-p2.pt')  # Assuming 'yolov8-p2.pt' is the pre-trained model's name
results = model.train(data='VisDrone.yaml', epochs=25, imgsz=224)
```

Good luck with your training on the VisDrone2019 dataset! 🤖

**Comment by bshakya77** [comment]
Thank you for your feedback. I will take a look at it.

Regards,
Bijay

**Comment by glenn-jocher** [comment]
@bshakya77 you're welcome, Bijay! If you have any more questions, feel free to reach out. Happy to help! 🚀

**Comment by fatbringer** [comment]
@glenn-jocher  i am back after doing some training and combining 2 datasets together that fits my use case better (Visdrone and MOT-Det)

I did training and my graphs look quite perculiar
![results](https://github.com/ultralytics/ultralytics/assets/93758179/9fabf497-004d-4856-b26b-0d1c06fdff1c)

I used 300 epochs, batch size of 16, imgsz of 640. It had went to full 300 epochs before stopping. There was something strange i noticed about the training losses, occuring at near 300 epochs, a sudden drop. Should this be a source for concern? 
Not too sure if it means it is overfitting or not.

**Comment by glenn-jocher** [comment]
@fatbringer The sudden drop in training losses near the end of your training process is unusual, but not necessarily a cause for concern. It could be due to a number of factors, such as a change in learning rate or the model finally converging on a particularly difficult aspect of the dataset. 

To determine if overfitting is occurring, compare your training and validation losses. If the validation loss is not decreasing alongside the training loss, or if it starts to increase, that could indicate overfitting. Additionally, monitor your validation metrics (like mAP) to ensure they continue to improve or at least remain stable as training progresses.

If you're satisfied with the detection performance and the validation metrics are strong, the model is likely performing well despite the peculiar loss graph. Keep an eye on these indicators for future training runs. 📈👀

**Comment by fatbringer** [comment]
@glenn-jocher from my graphs, the validation losses and mAP seems to have already plateaued, although the training loss has decreased. Should that ring any alarm bells? 

For learning rate, I had used according to the default which from my understanding, means it decreases linearly along with the epochs. I started with a lr of 0.02. (I also started with a lr of 0.02 when training yolov5 in the past, for comparing the 2 models). 

Considering that the model may only converge so late, is there any benefit to using a larger learning rate? 

Or would the other method of using cosine annealing be more helpful?

**Comment by glenn-jocher** [comment]
@fatbringer If your validation losses and mAP have plateaued while training loss has decreased, it's a sign that the model may not be improving on the validation set anymore. This isn't necessarily alarming, but it does suggest that the model might have reached its learning capacity with the current settings.

Using a larger learning rate could potentially help the model converge faster, but it also risks overshooting the optimal weights. On the other hand, cosine annealing is a learning rate schedule that can lead to better convergence by adjusting the learning rate in a non-linear fashion, potentially avoiding local minima.

You might want to experiment with both approaches to see which yields better results for your specific use case. Remember to monitor both training and validation metrics to gauge the effectiveness of these changes. 🔄📊

**Comment by fatbringer** [comment]
@glenn-jocher thanks for the ideas! I have tried the cosine annealing by setting cos_lr = true. The mAP values are quite similar to when using the linear learning rate

Just to check with you,  from my ClearML monitor, I noticed that the lr progresses like that
![image](https://github.com/ultralytics/ultralytics/assets/93758179/6623b35e-94ba-4ea0-abca-71f5c7764a20)

instead of something like that
![image](https://github.com/ultralytics/ultralytics/assets/93758179/9ed34ba1-e4fe-4e26-9c6b-42a22eb2165f)


would it be better with the restarts?
Does yolov8 training support the method with restarts?

**Comment by glenn-jocher** [comment]
@fatbringer Great to hear that cosine annealing is showing improvements on your test set!

The first learning rate graph you're seeing is a typical cosine annealing schedule without restarts. The second graph, which includes periodic restarts, is known as cosine annealing with warm restarts (or SGDR). YOLOv8 training does support various learning rate schedules, including cosine annealing with or without restarts. You can experiment with restarts to see if it benefits your specific scenario.

Regarding training from scratch versus starting with pre-trained weights, it often depends on how similar your target domain is to the data the model was pre-trained on. If the drone footage is significantly different from the pre-trained dataset, starting from scratch might be beneficial. However, pre-trained weights can still provide a useful starting point and often lead to faster convergence, especially if you have a limited amount of training data. 🚁🤖

**Comment by glenn-jocher** [comment]
@fatbringer to train from scratch, simply instantiate the model without specifying a pre-trained checkpoint:

```python
model = YOLO()  # Train from scratch
```

For cosine annealing with restarts, set the `cos_lr` parameter to `True` and adjust the `epochs` parameter to define the restart interval:

```python
results = model.train(data='your_dataset.yaml', epochs=300, imgsz=640, cos_lr=True)
```

The `epochs` parameter will determine the length of each cycle before the learning rate is reset. 🔄📈

**Comment by fatbringer** [comment]
@glenn-jocher Got it. Thanks for your patient explanation! 

How might I specify the model size? Such as to use YOLOv8s or YOLOv8m ?

**Comment by glenn-jocher** [comment]
@fatbringer You're welcome! To specify the model size, you can pass the model configuration file name when creating the YOLO object. For example:

```python
model = YOLO('yolov8s.yaml')  # for YOLOv8 small
model = YOLO('yolov8m.yaml')  # for YOLOv8 medium
```

Choose the appropriate configuration file for the model size you want to train. Happy training! 🚀

**Comment by fatbringer** [comment]
@glenn-jocher thanks!

I have another question. How does the training with the cosine annealing restarts know when to stop? 

Now I have set epochs=20 and cos_lr = true. The training terminal shows me 1/20 epochs. How do I set it such that it goes to 300 epochs in total, but resets every say 20 or 50 epochs?

**Comment by glenn-jocher** [comment]
@fatbringer To achieve cosine annealing with restarts over 300 epochs with resets every 20 or 50 epochs, you would need to implement a custom learning rate scheduler or modify the existing one, as the current implementation does not directly support specifying a reset interval. Set `epochs=300` for the total duration and manage the learning rate schedule within your training loop or by modifying the scheduler in the source code. 🔄🕒

**Comment by glenn-jocher** [comment]
@fatbringer In YOLOv8, the model automatically switches to evaluation mode during the validation process, so you don't need to explicitly call `model.eval()`. If you're experiencing unexpected behavior, ensure you're using the latest version of the repository and check the documentation for updates on the validation process. 📚🔍

**Comment by ahmedfarouk99** [comment]
@fatbringer did you managed to customize the source code to " Now I have set epochs=20 and cos_lr = true. The training terminal shows me 1/20 epochs. The training terminated at 20 epochs.
How do I set it such that it goes to 300 epochs in total, but the learning rate resets every say 20 or 50 epochs? "
if so, how you did it?
thanks in advance

**Comment by fatbringer** [comment]
@ahmedfarouk99 hey there. No unfortunately i was not able to do that. I suppose to the closest method to that will be re-executing the training code from the last saved checkpoint after every 20 epoch. Very manual unfortunately.

I wonder if it is possible to write a python script to call that rain script? Or probably use some bash or shell script, but i am not familiar. Cant be of help there.
